### PR TITLE
feat: Add PriorityQueue

### PR DESCRIPTION
**Pull request by Andrei1998**
This PR adds PriorityQueue support to the core library, implementing two variants:

1. **Binary heap-based implementation** using `List` (under `src/PriorityQueue.mo`).
2. **Wrapper over `Set`** (under `bench/utils/PriorityQueueSet.mo` and intended for benchmarking/testing purposes only).

Resolves https://github.com/dfinity/motoko-core/issues/391.

# Tests

The PR contains comprehensive unit tests, including cross-testing the two implementations against each other to ensure correctness, as well as more complex test scenarios such as long randomized sequences of operations and exhaustive testing of all possible small sequences to catch edge cases.

# Benchmarks

The PR contains extensive benchmarks comparing the performance of both implementations (on `Nat` values). See table below.

_Note:_ an earlier commit in this PR considered simpler but less efficient ways of implementing `push` and `pop` for the heap-based implementation. These were removed in favor of more efficient implementations based on the gap technique to avoid needless copies. Concretely, commit 664d09253e4c03a319388709f520f0a8ac44e6ac distinguished between:
- `push` (removed) and `pushBetter` (now renamed `push`).
- `pop` (removed) and `popBetter` (now renamed `pop`).

The table below is for the state of the code at this commit.

--------

Instructions

|                                           | A) PriorityQueue | A2) PriorityQueue better push | A3) PriorityQueue better push, better pop | B) PriorityQueueSet |
| :---------------------------------------- | ---------------: | ----------------------------: | ----------------------------------------: | ------------------: |
| 1.) 100000 operations (push:pop = 1:1)    |    1_078_836_395 |                   871_391_356 |                               598_903_656 |         522_729_799 |
| 2.) 100000 operations (push:pop = 2:1)    |    1_365_419_609 |                 1_081_007_251 |                               744_336_786 |         809_693_353 |
| 3.) 100000 operations (push:pop = 10:1)   |      598_673_615 |                   461_615_949 |                               359_307_559 |         873_181_417 |
| 4.) 100000 operations (only push)         |      271_131_449 |                   193_823_567 |                               193_823_107 |         886_825_181 |
| 5.) 50000 pushes, then 50000 pops         |    1_347_356_950 |                 1_308_569_002 |                               778_032_800 |         961_777_046 |
| 6.) 50000 pushes, then 25000 "pop;push"es |      902_261_756 |                   813_912_597 |                               530_870_481 |         922_137_746 |

Heap (@alexandru-uta noted that these seem broken at the moment, likely due to independent reasons):

|                                           | A) PriorityQueue | A2) PriorityQueue better push | A3) PriorityQueue better push, better pop | B) PriorityQueueSet |
| :---------------------------------------- | ---------------: | ----------------------------: | ----------------------------------------: | ------------------: |
| 1.) 100000 operations (push:pop = 1:1)    |            308 B |                         272 B |                                     272 B |               272 B |
| 2.) 100000 operations (push:pop = 2:1)    |            308 B |                         308 B |                                     272 B |               272 B |
| 3.) 100000 operations (push:pop = 10:1)   |            272 B |                         272 B |                                     272 B |               272 B |
| 4.) 100000 operations (only push)         |            272 B |                         272 B |                                     272 B |               272 B |
| 5.) 50000 pushes, then 50000 pops         |            308 B |                         308 B |                                     272 B |               272 B |
| 6.) 50000 pushes, then 25000 "pop;push"es |            272 B |                         272 B |                                     272 B |               272 B |

Garbage Collection

|                                           | A) PriorityQueue | A2) PriorityQueue better push | A3) PriorityQueue better push, better pop | B) PriorityQueueSet |
| :---------------------------------------- | ---------------: | ----------------------------: | ----------------------------------------: | ------------------: |
| 1.) 100000 operations (push:pop = 1:1)    |        10.23 MiB |                     15.03 MiB |                                 15.03 MiB |           17.43 MiB |
| 2.) 100000 operations (push:pop = 2:1)    |        12.91 MiB |                     19.73 MiB |                                 19.73 MiB |           19.32 MiB |
| 3.) 100000 operations (push:pop = 10:1)   |         4.22 MiB |                      8.67 MiB |                                  8.67 MiB |           12.64 MiB |
| 4.) 100000 operations (only push)         |       406.75 KiB |                      3.87 MiB |                                  3.87 MiB |            9.96 MiB |
| 5.) 50000 pushes, then 50000 pops         |        20.29 MiB |                     22.03 MiB |                                 22.03 MiB |            26.2 MiB |
| 6.) 50000 pushes, then 25000 "pop;push"es |        11.05 MiB |                     14.22 MiB |                                 14.22 MiB |           18.44 MiB |

### CODE REVIEW

File: src/PriorityQueue.mo [hunk 1]
```diff
@@ -0,0 +1,288 @@
...
+          index := (index - 1) / 2
+        };
+        case _ return
+      }
+    }
+  };
+
+  // Optimized version.
+  // Main optimization is using the hole technique to avoid copies.
+  public func pushBetter<T>(
```
[the remark this answers is not in the document]
Comment (Andrei1998) [reply src/PriorityQueue.mo:153]: For sure! This was also my intention, as in the PR description:

> **Heap-based implementation** using `List`
> 
> * Includes optimized variants `pushBetter` and `popBetter`. These are less readable but, as shown in the benchmarks, provide a significant performance improvement.
> * _Note: We will not merge the standard `push` and `pop` without the “Better” optimizations. They are included here (temporarily renamed) only to facilitate benchmarking._

Was thinking to do this more or less right before merging, but can also do otherwise.

[the remark this answers is not in the document]
Comment (rvanasa) [reply src/PriorityQueue.mo:153]: Missed that; thanks. We can see if anyone else has feedback on the implementation. Otherwise, LGTM!

**Review by rvanasa — APPROVED** [review]

**Comment by rvanasa** [comment]
Should be ready to merge after updating the `better*` functions. Thanks again!

File: src/internal/PriorityQueueSet.mo [hunk 2]
```diff
@@ -0,0 +1,156 @@
+/// A mutable priority queue of elements.
```
[the remark this answers is not in the document]
Comment (Andrei1998) [reply src/internal/PriorityQueueSet.mo:1]: Good point, should do that! However, I do use it in tests as well (so it's used in benchmarks and tests), any suggestion for the file structure then? I could put it under tests/utils or under bench/utils, but neither is much better than the other one I guess. This might have been the reason I tried putting it into src/internal back when I wrote it.

[the remark this answers is not in the document]
Comment (rvanasa) [reply src/internal/PriorityQueueSet.mo:1]: Got it; that makes sense. I'd say maybe just include it in `bench/utils` for now, since the primary purpose is to compare the performance between the two data structures. We can easily change this in the future as long as it's outside of the `src` directory (otherwise technically a breaking change).
